pkg/agent/ipassigner/responder/arp_responder.go

```diff
@@ -0,0 +1,141 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package responder
+
+import (
+	"fmt"
+	"net"
+	"sync"
+
+	"github.com/mdlayher/arp"
+	"github.com/mdlayher/ethernet"
+	"k8s.io/klog/v2"
+)
+
+type arpResponder struct {
+	iface       *net.Interface
+	conn        *arp.Client
+	assignedIPs map[string]struct{}
+	mutex       sync.Mutex
+	closed      chan struct{}
+}
+
+var _ Responder = (*arpResponder)(nil)
+
+func NewARPResponder(iface *net.Interface) (*arpResponder, error) {
+	conn, err := arp.Dial(iface)
+	if err != nil {
+		return nil, fmt.Errorf("creating ARP responder for %q: %s", iface.Name, err)
+	}
+	return &arpResponder{
+		iface:       iface,
+		conn:        conn,
+		assignedIPs: make(map[string]struct{}),
+		closed:      make(chan struct{}),
+	}, nil
+}
+
+// Advertise sends an gratuitous ARP packet for the IP.
+func (r *arpResponder) Advertise(ip net.IP) error {
+	select {
+	case <-r.closed:
+		return fmt.Errorf("ARP responder socket is closed")
+	default:
+		if ip.To4() == nil {
+			return fmt.Errorf("only IPv4 is supported")
+		}
+		pkt, err := arp.NewPacket(arp.OperationRequest, r.iface.HardwareAddr, ip, ethernet.Broadcast, ip)
+		if err != nil {
+			return err
+		}
+		return r.conn.WriteTo(pkt, ethernet.Broadcast)
+	}
+}
+
+func (r *arpResponder) InterfaceName() string {
+	return r.iface.Name
+}
+
+func (r *arpResponder) AddIP(ip net.IP) error {
+	if ip.To4() == nil {
+		return fmt.Errorf("only IPv4 is supported")
+	}
+	r.addIP(ip)
+	klog.InfoS("Assigned IP to ARP responder", "ip", ip, "interface", r.iface.Name)
+	return nil
+}
+
+func (r *arpResponder) RemoveIP(ip net.IP) error {
+	if ip.To4() == nil {
+		return fmt.Errorf("only IPv4 is supported")
+	}
+	r.deleteIP(ip)
+	klog.InfoS("Removed IP from ARP responder", "ip", ip, "interface", r.iface.Name)
+	return nil
+}
+
+func (r *arpResponder) handleARPRequest() error {
+	pkt, _, err := r.conn.Read()
+	if err != nil {
+		return err
+	}
+	if pkt.Operation != arp.OperationRequest {
+		return nil
+	}
+	if !r.isIPAssigned(pkt.TargetIP) {
+		klog.V(4).InfoS("Ignore ARP request", "ip", pkt.TargetIP, "interface", r.iface.Name)
+		return nil
+	}
+	klog.V(4).InfoS("Send ARP response", "ip", pkt.TargetIP, "interface", r.iface.Name)
+	if err := r.conn.Reply(pkt, r.iface.HardwareAddr, pkt.TargetIP); err != nil {
+		return fmt.Errorf("failed to reply ARP packet for IP %s: %v", pkt.TargetIP, err)
+	}
+	return nil
+}
+
+func (r *arpResponder) Run(stopCh <-chan struct{}) {
+	for {
+		select {
+		case <-stopCh:
+			close(r.closed)
+			r.conn.Close()
+			return
+		default:
+			err := r.handleARPRequest()
+			if err != nil {
+				klog.ErrorS(err, "Handle ARP request error", "deviceName", r.iface.Name)
```

Comment:
"Handle ARP request error" sounds it's handling error. It could be:

- Handling ARP request failed
- Failed to handle ARP request
- Error handling ARP request